{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Whole Genome Sequenced Mycobacterium bovis analysis for BTBgenie.\n",
    "\n",
    "Note: the current version of the M.bovis genome online has genes missing in the annotation. See https://www.ncbi.nlm.nih.gov/nuccore/LT708304.1?report=graph\n",
    "\n",
    "Updated files can be found here: https://github.com/dmnfarrell/gordon-group/tree/master/mbovis_annotation\n",
    "\n",
    "links:\n",
    "\n",
    "* https://github.com/JosephCrispell/GeneralTools/tree/master/ProcessingPipeline\n",
    "* [Woodchester Park eLife paper](https://elifesciences.org/articles/45833)\n",
    "* [An African origin for Mycobacterium bovis](https://academic.oup.com/emph/article/2020/1/49/5719036)\n",
    "* http://www.htslib.org/doc/bcftools.html\n",
    "* https://mtbgenomicsworkshop.readthedocs.io/\n",
    "* [BCFtools cheat sheet](https://gist.github.com/elowy01/93922762e131d7abd3c7e8e166a74a0b)\n",
    "* https://biopython.org/wiki/Category:Cookbook\n",
    "* https://pyvcf.readthedocs.io/en/v0.4.6/INTRO.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os,shutil,subprocess\n",
    "import random\n",
    "import glob, time\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 200)\n",
    "import pylab as plt\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from gzip import open as gzopen\n",
    "sys.path.append('pathogenie')\n",
    "from pathogenie import tools, aligners, app, trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mbovis samples from SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra = pd.read_csv('../sra_mbovis_runtable.csv')\n",
    "#sra.geo_loc_name_country.value_counts()\n",
    "#sub=sra[(sra.geo_loc_name_country=='Ireland') & (sra.Host=='COW') ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get master table with Loiseau et al. meta dataÂ¶"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('../Loiseau_TableS1.xlsx')\n",
    "master = pd.merge(df,sra,left_on='BIOSAMPLE',right_on='BioSample',how='inner')\n",
    "master.to_csv('../mbovis_sra_master.csv',index=False)\n",
    "cols = ['Run','CLONAL_COMPLEX','COUNTRY_ISOLATION','LibraryLayout','Host','ReleaseDate','Bytes','Bases','GENOME_COVERAGE']\n",
    "master.columns\n",
    "#master.geo_loc_name_country.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Representative samples of lineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Run CLONAL_COMPLEX COUNTRY_ISOLATION LibraryLayout        Host           ReleaseDate        Bytes        Bases  GENOME_COVERAGE\n",
      "36     ERR841808            Eu1                UK        PAIRED         Cow  2015-12-02T00:00:00Z  130651179.0  196040478.0            39.78\n",
      "245   SRR5216872            Eu1       New Zealand        PAIRED      BOVINE  2017-01-31T00:00:00Z  194523112.0  295803584.0            41.39\n",
      "464   SRR5486071       Unknown2          Ethiopia        PAIRED  Bos taurus  2017-04-28T00:00:00Z  251875435.0  445066100.0            50.71\n",
      "677    ERR125601            Eu1           Ireland        PAIRED         NaN  2012-10-01T00:00:00Z  277182504.0  346340680.0            68.83\n",
      "765   ERR2815558            Af1               NaN        PAIRED         NaN  2018-09-29T00:00:00Z  162711425.0  367702705.0            67.64\n",
      "1118  SRR1792002            Eu2               USA        PAIRED      Cattle  2015-03-11T00:00:00Z  155793754.0  221710448.0            37.69\n",
      "2577  SRR8065079            Af2            Uganda        PAIRED  Bos taurus  2018-12-01T00:00:00Z  149270806.0  248725188.0            32.01\n",
      "2597  SRR8063654            Af2          Ethiopia        PAIRED  Bos taurus  2018-12-01T00:00:00Z  128374182.0  220964011.0            25.87\n",
      "2608  SRR8063665            Af2          Ethiopia        PAIRED  Bos taurus  2018-12-01T00:00:00Z  127610605.0  219741248.0            27.41\n"
     ]
    }
   ],
   "source": [
    "sub=master[(master.LibraryLayout=='PAIRED') & (master.Bases<4e8) & (master.COUNTRY_ISOLATION.notnull())]\n",
    "res=[]\n",
    "lineages = ['Eu1','Eu2','Af2']\n",
    "for lin in lineages:\n",
    "    s = sub[(sub.CLONAL_COMPLEX==lin)].sample(2,random_state=4,replace=True)    \n",
    "    res.append(s)\n",
    "    \n",
    "accessions=['ERR2815558','SRR1792002','SRR5216872','SRR5486071','ERR841808','ERR125601',\n",
    "            'SRR8063654','SRR8063665','SRR8065079']\n",
    "testset = master[master.Run.isin(accessions)]\n",
    "print (testset[cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fastq-dump --split-files ERR125601 --outdir /storage/btbgenie/mbovis_sra\n"
     ]
    }
   ],
   "source": [
    "# fetch from SRA\n",
    "def fetch_sra_reads(df,path):\n",
    "    \"\"\"download a subset of reads \"\"\"\n",
    "    \n",
    "    for i,r in df.iterrows():\n",
    "        files = glob.glob(os.path.join(path,r.Run+'*'))        \n",
    "        if len(files)==0:\n",
    "            cmd = 'fastq-dump --split-files {n} --outdir {o}'.format(n=r.Run,o=path)\n",
    "            print (cmd)\n",
    "            subprocess.check_output(cmd,shell=True)\n",
    "\n",
    "fetch_sra_reads(testset,path='/storage/btbgenie/mbovis_sra')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blasting to known contaminants\n",
    "\n",
    "* https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2684-x#Sec2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f 4 mapped/3_S57_L001.bam > unmapped.bam\n"
     ]
    }
   ],
   "source": [
    "#get unmapped reads from a bam file\n",
    "\n",
    "infile='mapped/3_S57_L001.bam'\n",
    "cmd = 'samtools view -b -f 4 {i} > unmapped.bam'.format(i=infile)\n",
    "print (cmd)\n",
    "#align with bwa to known index of contaminants\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examine FASTQ file quality and define trimming parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f15d42d8750>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAARuUlEQVR4nO3db4xc1XnH8e9TXAjg1gacbqlt1W5jJaK4Sc0KiJCidRwlBqKYFyQishI7deU3JKHBVTGJVNRWbR21lILUUlmBYqqITeKmxYL8cw2rKC/sxs4fbHAoG2LAK2OHYNwukCZWn76Y42a17NqeGe+s957vR1rNveeeuec8uuY3M/feGSIzkSTV4ZemewKSpN4x9CWpIoa+JFXE0Jekihj6klSRWdM9gZOZN29eLlq0qKdjvvrqq1x44YU9HXMqNa0eaF5NTasHmlfTTKtnz549L2XmmyfadlaH/qJFi9i9e3dPxxwaGmJgYKCnY06lptUDzaupafVA82qaafVExHOTbfP0jiRVxNCXpIoY+pJUkVOGfkTcHxFHImLfmLaLI2J7RDxTHi8q7RER90TEcEQ8ERHLxjxnTen/TESsmZpyJEknczrv9B8AVo5r2wjsyMwlwI6yDnAtsKT8rQfuhdaLBHAHcBVwJXDHiRcKSVLvnDL0M/ObwMvjmlcBW8ryFuCGMe0PZstOYG5EXAq8D9iemS9n5lFgO298IZEkTbFOb9nsy8xDZflFoK8szwdeGNPvYGmbrP0NImI9rU8J9PX1MTQ01OEUOzM6OtrzMadS0+qB5tXUtHqgeTU1qZ6u79PPzIyIM/b7zJm5GdgM0N/fn72+N3am3Y97Kk2rB5pXU9PqgebV1KR6Or1753A5bUN5PFLaR4CFY/otKG2TtUuSeqjTd/rbgDXApvL48Jj2j0fEIK2Ltscy81BEfB34yzEXb98L3N75tCVp6i3a+CgAG5YeZ21Z7pUDm66fkv2eMvQj4iFgAJgXEQdp3YWzCfhiRKwDngM+VLp/BbgOGAZeAz4GkJkvR8SfA98u/f4sM8dfHJYkTbFThn5mfniSTSsm6JvAzZPs537g/rZmJ0k6o/xGriRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVaSr0I+IT0XEkxGxLyIeiog3RcTiiNgVEcMR8YWIOLf0Pa+sD5fti85EAZKk09dx6EfEfOCTQH9mXg6cA9wEfBa4KzPfAhwF1pWnrAOOlva7Sj9JUg91e3pnFnB+RMwCLgAOAe8GtpbtW4AbyvKqsk7ZviIiosvxJUltiMzs/MkRtwB/AbwOfAO4BdhZ3s0TEQuBr2bm5RGxD1iZmQfLth8CV2XmS+P2uR5YD9DX13fF4OBgx/PrxOjoKLNnz+7pmFOpafVA82pqWj3QnJr2jhwDoO98OPx6b8deOn9Ox89dvnz5nszsn2jbrE53GhEX0Xr3vhh4BfgSsLLT/Z2QmZuBzQD9/f05MDDQ7S7bMjQ0RK/HnEpNqweaV1PT6oHm1LR246MAbFh6nDv3dhyXHTmwemBK9tvN6Z33AD/KzB9n5s+BLwPXAHPL6R6ABcBIWR4BFgKU7XOAn3QxviSpTd2E/vPA1RFxQTk3vwJ4CngcuLH0WQM8XJa3lXXK9seym3NLkqS2dRz6mbmL1gXZ7wB7y742A7cBt0bEMHAJcF95yn3AJaX9VmBjF/OWJHWgq5NUmXkHcMe45meBKyfo+1Pgg92MJ0nqjt/IlaSKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0JekinQV+hExNyK2RsQPImJ/RLwzIi6OiO0R8Ux5vKj0jYi4JyKGI+KJiFh2ZkqQJJ2ubt/p3w18LTPfBrwd2A9sBHZk5hJgR1kHuBZYUv7WA/d2ObYkqU0dh35EzAHeBdwHkJk/y8xXgFXAltJtC3BDWV4FPJgtO4G5EXFpxzOXJLUtMrOzJ0a8A9gMPEXrXf4e4BZgJDPnlj4BHM3MuRHxCLApM79Vtu0AbsvM3eP2u57WJwH6+vquGBwc7Gh+nRodHWX27Nk9HXMqNa0eaF5NTasHmlPT3pFjAPSdD4df7+3YS+fP6fi5y5cv35OZ/RNtm9XxXlvPXQZ8IjN3RcTd/OJUDgCZmRHR1qtKZm6m9WJCf39/DgwMdDHF9g0NDdHrMadS0+qB5tXUtHqgOTWt3fgoABuWHufOvd3EZfsOrB6Ykv12c07/IHAwM3eV9a20XgQOnzhtUx6PlO0jwMIxz19Q2iRJPdJx6Gfmi8ALEfHW0rSC1qmebcCa0rYGeLgsbwM+Wu7iuRo4lpmHOh1fktS+bj+vfAL4fEScCzwLfIzWC8kXI2Id8BzwodL3K8B1wDDwWukrSeqhrkI/M78HTHSxYMUEfRO4uZvxJEnd8Ru5klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUka5DPyLOiYjvRsQjZX1xROyKiOGI+EJEnFvazyvrw2X7om7HliS150y8078F2D9m/bPAXZn5FuAosK60rwOOlva7Sj9JUg91FfoRsQC4HvhcWQ/g3cDW0mULcENZXlXWKdtXlP6SpB6JzOz8yRFbgb8CfgX4I2AtsLO8myciFgJfzczLI2IfsDIzD5ZtPwSuysyXxu1zPbAeoK+v74rBwcGO59eJ0dFRZs+e3dMxp1LT6oHm1dS0eqA5Ne0dOQZA3/lw+PXejr10/pyOn7t8+fI9mdk/0bZZne40It4PHMnMPREx0Ol+xsvMzcBmgP7+/hwYOGO7Pi1DQ0P0esyp1LR6oHk1Na0eaE5Nazc+CsCGpce5c2/HcdmRA6sHpmS/3VRxDfCBiLgOeBPwq8DdwNyImJWZx4EFwEjpPwIsBA5GxCxgDvCTLsaXJLWp43P6mXl7Zi7IzEXATcBjmbkaeBy4sXRbAzxclreVdcr2x7Kbc0uSpLZNxX36twG3RsQwcAlwX2m/D7iktN8KbJyCsSVJJ3FGTlJl5hAwVJafBa6coM9PgQ+eifEkSZ3xG7mSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klSRjkM/IhZGxOMR8VREPBkRt5T2iyNie0Q8Ux4vKu0REfdExHBEPBERy85UEZKk09PNO/3jwIbMvAy4Grg5Ii4DNgI7MnMJsKOsA1wLLCl/64F7uxhbktSBjkM/Mw9l5nfK8n8D+4H5wCpgS+m2BbihLK8CHsyWncDciLi045lLktoWmdn9TiIWAd8ELgeez8y5pT2Ao5k5NyIeATZl5rfKth3AbZm5e9y+1tP6JEBfX98Vg4ODXc+vHaOjo8yePbunY06lptUDzaupafVAc2raO3IMgL7z4fDrvR176fw5HT93+fLlezKzf6JtszreaxERs4F/Af4wM/+rlfMtmZkR0darSmZuBjYD9Pf358DAQLdTbMvQ0BC9HnMqNa0eaF5NTasHmlPT2o2PArBh6XHu3Nt1XLblwOqBKdlvV3fvRMQv0wr8z2fml0vz4ROnbcrjkdI+Aiwc8/QFpU2S1CPd3L0TwH3A/sz82zGbtgFryvIa4OEx7R8td/FcDRzLzEOdji9Jal83n1euAT4C7I2I75W2TwObgC9GxDrgOeBDZdtXgOuAYeA14GNdjC1J6kDHoV8uyMYkm1dM0D+BmzsdT5LUPb+RK0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFel56EfEyoh4OiKGI2Jjr8eXpJr1NPQj4hzg74FrgcuAD0fEZb2cgyTVbFaPx7sSGM7MZwEiYhBYBTzV43moIRZtfHS6p9C1DUuPs7aNOg5sun4KZ3N2asJxPltEZvZusIgbgZWZ+Qdl/SPAVZn58TF91gPry+pbgad7NsGWecBLPR5zKjWtHmheTU2rB5pX00yr5zcz880Tbej1O/1TyszNwObpGj8idmdm/3SNf6Y1rR5oXk1NqweaV1OT6un1hdwRYOGY9QWlTZLUA70O/W8DSyJicUScC9wEbOvxHCSpWj09vZOZxyPi48DXgXOA+zPzyV7O4TRM26mlKdK0eqB5NTWtHmheTY2pp6cXciVJ08tv5EpSRQx9SapI9aEfEedExHcj4pGyvjgidpWfifhCueA8Y0TE3IjYGhE/iIj9EfHOiLg4IrZHxDPl8aLpnufpiohPRcSTEbEvIh6KiDfNtGMUEfdHxJGI2DembcJjEi33lNqeiIhl0zfziU1Sz1+Xf3NPRMS/RsTcMdtuL/U8HRHvm55Zn9xENY3ZtiEiMiLmlfWz/hidTPWhD9wC7B+z/lngrsx8C3AUWDcts+rc3cDXMvNtwNtp1bYR2JGZS4AdZf2sFxHzgU8C/Zl5Oa2L/zcx847RA8DKcW2THZNrgSXlbz1wb4/m2I4HeGM924HLM/N3gf8EbgcoP7NyE/A75Tn/UH6O5WzzAG+siYhYCLwXeH5M80w4RpOqOvQjYgFwPfC5sh7Au4GtpcsW4IbpmV37ImIO8C7gPoDM/FlmvkLrpy62lG4zqiZad5idHxGzgAuAQ8ywY5SZ3wReHtc82TFZBTyYLTuBuRFxaW9menomqiczv5GZx8vqTlrfwYFWPYOZ+T+Z+SNgmNbPsZxVJjlGAHcBfwyMvePlrD9GJ1N16AN/R+uA/m9ZvwR4Zcw/3oPA/OmYWIcWAz8G/qmcsvpcRFwI9GXmodLnRaBv2mbYhswcAf6G1rusQ8AxYA8z+xidMNkxmQ+8MKbfTKzv94GvluUZW09ErAJGMvP74zbN2Jqg4tCPiPcDRzJzz3TP5QyaBSwD7s3M3wNeZdypnGzdozsj7tMt57lX0Xox+w3gQib4CD7TzaRjcioR8RngOPD56Z5LNyLiAuDTwJ9M91zOtGpDH7gG+EBEHAAGaZ0yuJvWR7UTX1qbaT8TcRA4mJm7yvpWWi8Ch098/CyPR6Zpfu16D/CjzPxxZv4c+DKt4zaTj9EJkx2TGftTJRGxFng/sDp/8QWgmVrPb9N6s/H9khELgO9ExK8zc2sCKg79zLw9Mxdk5iJaF5oey8zVwOPAjaXbGuDhaZpi2zLzReCFiHhraVpB62ert9GqBWZWTc8DV0fEBeV6y4l6ZuwxGmOyY7IN+Gi5Q+Rq4NiY00BnrYhYSetU6Qcy87Uxm7YBN0XEeRGxmNbFz/+Yjjm2IzP3ZuavZeaikhEHgWXlv7EZeYz+X2ZW/wcMAI+U5d+i9Y9yGPgScN50z6/NWt4B7AaeAP4NuIjWtYodwDPAvwMXT/c826jnT4EfAPuAfwbOm2nHCHiI1jWJn9MKj3WTHRMgaP2Phn4I7KV159K013Aa9QzTOs/9vfL3j2P6f6bU8zRw7XTP/3RrGrf9ADBvphyjk/35MwySVJFqT+9IUo0MfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klSR/wNp9udsdM8UwgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "reload(tools)\n",
    "\n",
    "df = tools.fastq_to_dataframe(testfile)\n",
    "df.length.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43.92692995071411\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f75432f2c50>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAATQElEQVR4nO3db4xcV3nH8e9Tm4TKm3rzh66s2KpNsUAREcFZJUYgtIsFddwKuxJEQRWxI1fuiwQFEaSYIrVUqtTQqkSJiqJumygOoixRIIqVBNrUZIXywgE7BDt/SLMJifDK2EpwDJsIaOjTF3MWNsuud2Z3Zoc98/1Io7n33HPnnsfX/u3dM3fGkZlIkurye90egCSp/Qx3SaqQ4S5JFTLcJalChrskVWhltwcAcMEFF+T69etb3u/VV19l1apV7R/Q7zBr7g29WDP0Zt2Lqfnw4cMvZeZbZtv2OxHu69ev59ChQy3vNzY2xtDQUPsH9DvMmntDL9YMvVn3YmqOiBfn2ua0jCRVyHCXpAoZ7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalChrskVWjeT6hGxNuBr05reivwN8BdpX098AJwZWaeiogAbgG2Aa8BuzLzsfYOW5LaZ/3eB7p27Du3dubrFua9cs/MZzLzksy8BLiURmDfC+wFDmTmRuBAWQe4AthYHnuA2zoxcEnS3FqdltkCPJeZLwLbgX2lfR+woyxvB+7KhoNAf0SsactoJUlNiVb+D9WIuAN4LDP/JSJeycz+0h7Aqczsj4j7gZsy85Gy7QBwY2YemvFae2hc2TMwMHDp6Ohoy4OfnJykr6+v5f2WM2vuDb1YM3Sv7qMTp5f8mFM2rF6x4JqHh4cPZ+bgbNua/lbIiDgL+DDwmZnbMjMjoqX/aTszR4ARgMHBwVzIt6L5DXK9wZp7R7fq3tXlOfdO1NzKtMwVNK7aT5T1E1PTLeX5ZGmfANZN229taZMkLZFWwv1jwFemre8HdpblncB909qvjobNwOnMPL7okUqSmtbUtExErAI+CPzVtOabgLsjYjfwInBlaX+Qxm2Q4zTurLmmbaOVJDWlqXDPzFeB82e0vUzj7pmZfRO4ti2jkyQtiJ9QlaQKGe6SVCHDXZIqZLhLUoUMd0mqkOEuSRUy3CWpQoa7JFXIcJekChnuklQhw12SKmS4S1KFDHdJqpDhLkkVMtwlqUKGuyRVyHCXpAoZ7pJUIcNdkipkuEtShZoK94joj4h7IuIHEfF0RLwnIs6LiIci4tnyfG7pGxFxa0SMR8SRiNjU2RIkSTM1e+V+C/DNzHwH8C7gaWAvcCAzNwIHyjrAFcDG8tgD3NbWEUuS5jVvuEfEauD9wO0AmfnLzHwF2A7sK932ATvK8nbgrmw4CPRHxJq2j1ySNKfIzDN3iLgEGAGeonHVfhi4HpjIzP7SJ4BTmdkfEfcDN2XmI2XbAeDGzDw043X30LiyZ2Bg4NLR0dGWBz85OUlfX1/L+y1n1twberFm6F7dRydOL/kxp2xYvWLBNQ8PDx/OzMHZtq1sYv+VwCbgE5n5aETcwm+mYADIzIyIM/+UmCEzR2j80GBwcDCHhoZa2R2AsbExFrLfcmbNvaEXa4bu1b1r7wNLfswpd25d1ZGam5lzPwYcy8xHy/o9NML+xNR0S3k+WbZPAOum7b+2tEmSlsi84Z6ZPwZ+FBFvL01baEzR7Ad2lradwH1leT9wdblrZjNwOjOPt3fYkqQzaWZaBuATwJcj4izgeeAaGj8Y7o6I3cCLwJWl74PANmAceK30lSQtoabCPTMfB2abtN8yS98Erl3kuCRJi+AnVCWpQoa7JFXIcJekChnuklQhw12SKmS4S1KFDHdJqpDhLkkVMtwlqUKGuyRVyHCXpAoZ7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalChrskVchwl6QKGe6SVKGmwj0iXoiIoxHxeEQcKm3nRcRDEfFseT63tEdE3BoR4xFxJCI2dbIASdJva+XKfTgzL8nMwbK+FziQmRuBA2Ud4ApgY3nsAW5r12AlSc1ZzLTMdmBfWd4H7JjWflc2HAT6I2LNIo4jSWpRZOb8nSJ+CJwCEvjXzByJiFcys79sD+BUZvZHxP3ATZn5SNl2ALgxMw/NeM09NK7sGRgYuHR0dLTlwU9OTtLX19fyfsuZNfeGXqwZulf30YnTS37MKRtWr1hwzcPDw4enzaa8wcomX+N9mTkREX8IPBQRP5i+MTMzIub/KfHGfUaAEYDBwcEcGhpqZXcAxsbGWMh+y5k194ZerBm6V/euvQ8s+TGn3Ll1VUdqbmpaJjMnyvNJ4F7gMuDE1HRLeT5Zuk8A66btvra0SZKWyLzhHhGrIuKcqWXgQ8ATwH5gZ+m2E7ivLO8Hri53zWwGTmfm8baPXJI0p2amZQaAexvT6qwE/iMzvxkR3wXujojdwIvAlaX/g8A2YBx4Dbim7aOWJJ3RvOGemc8D75ql/WVgyyztCVzbltFJkhbET6hKUoUMd0mqkOEuSRUy3CWpQoa7JFXIcJekChnuklQhw12SKmS4S1KFDHdJqpDhLkkVMtwlqUKGuyRVyHCXpAoZ7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalCTYd7RKyIiO9FxP1lfUNEPBoR4xHx1Yg4q7SfXdbHy/b1nRm6JGkurVy5Xw88PW3988DNmfk24BSwu7TvBk6V9ptLP0nSEmoq3CNiLfCnwL+X9QA+ANxTuuwDdpTl7WWdsn1L6S9JWiKRmfN3irgH+AfgHODTwC7gYLk6JyLWAd/IzHdGxBPA1sw8VrY9B1yemS/NeM09wB6AgYGBS0dHR1se/OTkJH19fS3vt5xZc2/oxZqhe3UfnTi95MecsmH1igXXPDw8fDgzB2fbtnK+nSPiz4CTmXk4IoYWNIJZZOYIMAIwODiYQ0Otv/TY2BgL2W85s+be0Is1Q/fq3rX3gSU/5pQ7t67qSM3zhjvwXuDDEbENeDPwB8AtQH9ErMzM14G1wETpPwGsA45FxEpgNfBy20cuSZrTvHPumfmZzFybmeuBq4BvZeZfAA8DHynddgL3leX9ZZ2y/VvZzNyPJKltFnOf+43ApyJiHDgfuL203w6cX9o/Bexd3BAlSa1qZlrm1zJzDBgry88Dl83S5+fAR9swNknSAvkJVUmqkOEuSRUy3CWpQoa7JFXIcJekChnuklQhw12SKmS4S1KFDHdJqpDhLkkVMtwlqUKGuyRVyHCXpAoZ7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalChrskVWjecI+IN0fEdyLi+xHxZET8XWnfEBGPRsR4RHw1Is4q7WeX9fGyfX1nS5AkzdTMlfsvgA9k5ruAS4CtEbEZ+Dxwc2a+DTgF7C79dwOnSvvNpZ8kaQnNG+7ZMFlW31QeCXwAuKe07wN2lOXtZZ2yfUtERNtGLEmaV2Tm/J0iVgCHgbcBXwT+CThYrs6JiHXANzLznRHxBLA1M4+Vbc8Bl2fmSzNecw+wB2BgYODS0dHRlgc/OTlJX19fy/stZ9bcG3qxZuhe3UcnTi/5MadsWL1iwTUPDw8fzszB2batbOYFMvNXwCUR0Q/cC7xjQSN542uOACMAg4ODOTQ01PJrjI2NsZD9ljNr7g29WDN0r+5dex9Y8mNOuXPrqo7U3NLdMpn5CvAw8B6gPyKmfjisBSbK8gSwDqBsXw283JbRSpKa0szdMm8pV+xExO8DHwSephHyHynddgL3leX9ZZ2y/VvZzNyPJKltmpmWWQPsK/PuvwfcnZn3R8RTwGhE/D3wPeD20v924EsRMQ78BLiqA+OWJJ3BvOGemUeAd8/S/jxw2SztPwc+2pbRSZIWxE+oSlKFDHdJqpDhLkkVMtwlqUKGuyRVyHCXpAoZ7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalChrskVchwl6QKGe6SVCHDXZIqZLhLUoUMd0mqkOEuSRUy3CWpQvOGe0Ssi4iHI+KpiHgyIq4v7edFxEMR8Wx5Pre0R0TcGhHjEXEkIjZ1ughJ0hs1c+X+OnBDZl4EbAaujYiLgL3AgczcCBwo6wBXABvLYw9wW9tHLUk6o3nDPTOPZ+ZjZflnwNPAhcB2YF/ptg/YUZa3A3dlw0GgPyLWtH3kkqQ5tTTnHhHrgXcDjwIDmXm8bPoxMFCWLwR+NG23Y6VNkrREVjbbMSL6gK8Bn8zMn0bEr7dlZkZEtnLgiNhDY9qGgYEBxsbGWtkdgMnJyQXtt5xZc2/oxZqhe3XfcPHrS37MKZ2qualwj4g30Qj2L2fm10vziYhYk5nHy7TLydI+Aaybtvva0vYGmTkCjAAMDg7m0NBQy4MfGxtjIfstZ9bcG3qxZuhe3bv2PrDkx5xy59ZVHam5mbtlArgdeDozvzBt035gZ1neCdw3rf3qctfMZuD0tOkbSdISaObK/b3Ax4GjEfF4aftr4Cbg7ojYDbwIXFm2PQhsA8aB14Br2jpiSdK85g33zHwEiDk2b5mlfwLXLnJckqRF8BOqklQhw12SKmS4S1KFDHdJqpDhLkkVMtwlqUKGuyRVyHCXpAoZ7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalChrskVchwl6QKGe6SVCHDXZIqZLhLUoUMd0mqkOEuSRWaN9wj4o6IOBkRT0xrOy8iHoqIZ8vzuaU9IuLWiBiPiCMRsamTg5ckza6ZK/c7ga0z2vYCBzJzI3CgrANcAWwsjz3Abe0ZpiSpFfOGe2Z+G/jJjObtwL6yvA/YMa39rmw4CPRHxJp2DVaS1JzIzPk7RawH7s/Md5b1VzKzvywHcCoz+yPifuCmzHykbDsA3JiZh2Z5zT00ru4ZGBi4dHR0tOXBT05O0tfX1/J+y5k194ZerBm6V/fRidNLfswpG1avWHDNw8PDhzNzcLZtKxc1KiAzMyLm/wnx2/uNACMAg4ODOTQ01PKxx8bGWMh+y5k194ZerBm6V/euvQ8s+TGn3Ll1VUdqXujdMiemplvK88nSPgGsm9ZvbWmTJC2hhV657wd2AjeV5/umtV8XEaPA5cDpzDy+6FFK6glHJ0539Sq6JvOGe0R8BRgCLoiIY8Df0gj1uyNiN/AicGXp/iCwDRgHXgOu6cCYJUnzmDfcM/Njc2zaMkvfBK5d7KAkSYvjJ1QlqUKGuyRVyHCXpAot+j53SXVZ38W7VW64uGuHro5X7pJUIcNdkipkuEtShQx3SaqQ4S5JFTLcJalC3gop/Y7q5i2JWv68cpekChnuklQhp2WkM+jW1MgNF7+O/zy1GF65S1KFDHdJqpDhLkkVMtwlqUK+Y6OWdOMNxhsuft3/NFlqkeG+DPnhFknz6ci0TERsjYhnImI8IvZ24hiSpLm1/co9IlYAXwQ+CBwDvhsR+zPzqXYfq9u6NUXhL1yS5tOJlLgMGM/M5wEiYhTYDnQk3J2ikKTfFpnZ3heM+AiwNTP/sqx/HLg8M6+b0W8PsKesvh14ZgGHuwB4aRHDXY6suTf0Ys3Qm3UvpuY/ysy3zLaha7/fZ+YIMLKY14iIQ5k52KYhLQvW3Bt6sWbozbo7VXMn3lCdANZNW19b2iRJS6QT4f5dYGNEbIiIs4CrgP0dOI4kaQ5tn5bJzNcj4jrgP4EVwB2Z+WS7j1MsalpnmbLm3tCLNUNv1t2Rmtv+hqokqfv8bhlJqpDhLkkVWrbh3itfcRARL0TE0Yh4PCIOlbbzIuKhiHi2PJ/b7XEuRkTcEREnI+KJaW2z1hgNt5bzfiQiNnVv5As3R82fi4iJcq4fj4ht07Z9ptT8TET8SXdGvTgRsS4iHo6IpyLiyYi4vrRXe67PUHPnz3VmLrsHjTdqnwPeCpwFfB+4qNvj6lCtLwAXzGj7R2BvWd4LfL7b41xkje8HNgFPzFcjsA34BhDAZuDRbo+/jTV/Dvj0LH0vKn/HzwY2lL/7K7pdwwJqXgNsKsvnAP9Taqv2XJ+h5o6f6+V65f7rrzjIzF8CU19x0Cu2A/vK8j5gRxfHsmiZ+W3gJzOa56pxO3BXNhwE+iNizdKMtH3mqHku24HRzPxFZv4QGKfxb2BZyczjmflYWf4Z8DRwIRWf6zPUPJe2nevlGu4XAj+atn6MM/+BLWcJ/FdEHC5f2QAwkJnHy/KPgYHuDK2j5qqx9nN/XZmCuGPadFt1NUfEeuDdwKP0yLmeUTN0+Fwv13DvJe/LzE3AFcC1EfH+6Ruz8btc1fez9kKNxW3AHwOXAMeBf+7ucDojIvqArwGfzMyfTt9W67mepeaOn+vlGu498xUHmTlRnk8C99L4Fe3E1K+n5flk90bYMXPVWO25z8wTmfmrzPw/4N/4za/j1dQcEW+iEXJfzsyvl+aqz/VsNS/FuV6u4d4TX3EQEasi4pypZeBDwBM0at1Zuu0E7uvOCDtqrhr3A1eXOyk2A6en/Uq/rM2YT/5zGucaGjVfFRFnR8QGYCPwnaUe32JFRAC3A09n5hembar2XM9V85Kc626/m7yId6G30Xjn+Tngs90eT4dqfCuNd86/Dzw5VSdwPnAAeBb4b+C8bo91kXV+hcavpv9LY45x91w10rhz4ovlvB8FBrs9/jbW/KVS05Hyj3zNtP6fLTU/A1zR7fEvsOb30ZhyOQI8Xh7baj7XZ6i54+farx+QpAot12kZSdIZGO6SVCHDXZIqZLhLUoUMd0mqkOEuSRUy3CWpQv8PhP2fOOoXoJYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "def trim_reads_default(filename, outfile, right_quality=30):\n",
    "    \"\"\"Trim adapters\"\"\"\n",
    "        \n",
    "    fastq_parser = SeqIO.parse(gzopen(filename, \"rt\"), \"fastq\")\n",
    "    c=0\n",
    "    out = gzopen(outfile, \"wt\")\n",
    "    for record in fastq_parser:\n",
    "        score = record.letter_annotations[\"phred_quality\"]         \n",
    "        for i in range(len(score)-1,0,-1):\n",
    "            if score[i] >= right_quality:\n",
    "                break\n",
    "        #trimmed.append(record[:i])\n",
    "        #print (record[:i])        \n",
    "        #c+=1\n",
    "        #if c>100:\n",
    "        #    break    \n",
    "        SeqIO.write(record[:i],out,'fastq')\n",
    "    return\n",
    "\n",
    "testfile = 'mbovis_sra/SRR1791711_1.fastq.gz'\n",
    "st=time.time()\n",
    "trim_reads_default(testfile,'trimmed.fastq.gz')\n",
    "print (time.time()-st)\n",
    "df = tools.fastq_to_dataframe('trimmed.fastq.gz')\n",
    "df.length.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### multiprocess trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 465388 records to group_1.fastq\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['group_1.fastq']"
      ]
     },
     "execution_count": 439,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def split_fastq(filename, chunks=4):\n",
    "    \"\"\"Split fastq file\"\"\"\n",
    "    \n",
    "    names=[]\n",
    "    tmp = '/tmp'\n",
    "    tmpfiles = []\n",
    "    record_iter = SeqIO.parse(gzopen(filename, \"rt\"), \"fastq\")\n",
    "    length = \n",
    "    for i, batch in enumerate(tools.batch_iterator(record_iter, chunksize)):\n",
    "        tempfile = \"group_%i.fastq\" % (i + 1)\n",
    "        tmpfiles.append(tempfile)\n",
    "        names.append(tempfile)\n",
    "        with open(tempfile, \"w\") as handle:\n",
    "            count = SeqIO.write(batch, handle, \"fastq\")\n",
    "        print(\"Wrote %i records to %s\" % (count, tempfile))\n",
    "    return names\n",
    "\n",
    "split_fastq(testfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools.plot_fastq_qualities(testfile,limit=100000)\n",
    "plt.savefig('fastq_quals.png',dpi=100)\n",
    "tools.plot_fastq_gc_content(testfile, limit=50000)\n",
    "plt.savefig('gc.png',dpi=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sample ids from fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path1='/storage/btbgenie/albania'\n",
    "files = glob.glob(os.path.join(path1,'*.fastq.gz'))\n",
    "path2='/storage/btbgenie/mbovis_sra'\n",
    "files = glob.glob(os.path.join(path2,'*.fastq.gz'))\n",
    "\n",
    "files = app.get_files_from_paths([path1,path2])\n",
    "\n",
    "df = app.get_samples(files)\n",
    "df\n",
    "#df.set_index(['sample','pair']).unstack()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tab file for multi snippy\n",
    "x = df.reset_index().pivot(columns='pair', index='sample', values='filename')\n",
    "x.to_csv('samples.tab',sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align the FASTQ files against reference and create VCF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(aligners)\n",
    "ref = app.ref_genome\n",
    "samples = app.get_samples(files)\n",
    "\n",
    "app.align_reads(samples, idx=ref, threads=10, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_samples():\n",
    "    import vcf\n",
    "    vcf_reader = vcf.Reader(open('mapped/calls.vcf.gz', 'rb'))\n",
    "    vcf_reader.samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### remove duplicated reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = 'samtools rmdup A_reads.bt2.sorted.bam A_reads.bt2.sorted.noDups.bam'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### call all samples at once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parallel bcftools mpileup -r {1} -O b -o {2} -f /home/damien/.config/pathogenie/genome/Mbovis_AF212297.fa mapped/SRR8063665.bam mapped/SRR1791711.bam mapped/SRR8063654.bam mapped/6_S7_L001.bam mapped/SRR5486071.bam mapped/3_S57_L001.bam ::: NC_002945.4:1-483322 NC_002945.4:483323-966645 NC_002945.4:966646-1449967 NC_002945.4:1449968-1933290 NC_002945.4:1933291-2416612 NC_002945.4:2416613-2899935 NC_002945.4:2899936-3383257 NC_002945.4:3383258-3866580 NC_002945.4:3866581-4349903 :::+ /tmp/1.bcf /tmp/483323.bcf /tmp/966646.bcf /tmp/1449968.bcf /tmp/1933291.bcf /tmp/2416613.bcf /tmp/2899936.bcf /tmp/3383258.bcf /tmp/3866581.bcf\n",
      "bcftools concat /tmp/1.bcf /tmp/483323.bcf /tmp/966646.bcf /tmp/1449968.bcf /tmp/1933291.bcf /tmp/2416613.bcf /tmp/2899936.bcf /tmp/3383258.bcf /tmp/3866581.bcf -O b -o mapped/raw.bcf\n",
      "bcftools call -V indels --ploidy 1 -m -v -o mapped/calls.vcf mapped/raw.bcf\n",
      "bcftools filter -e \"QUAL<40\" -o mapped/filtered.vcf.gz -O z mapped/calls.vcf\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def create_bam_labels(filenames):\n",
    "    \n",
    "    names = [os.path.basename(i).split('.')[0] for i in filenames]\n",
    "    #print (names)\n",
    "    with open('samples.txt','w+') as file:\n",
    "        for s in zip(bam_files,names):\n",
    "            file.write('%s %s\\n' %(s[0],s[1]))\n",
    "    return\n",
    "\n",
    "bam_files = glob.glob('mapped/*.bam')\n",
    "create_bam_labels(bam_files)\n",
    "vcf_file = app.variant_calling(bam_files, ref, 'mapped', threads=10, overwrite=False)#, sample_file='samples.txt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unknown    4342195\n",
       "indel        10772\n",
       "snp           7545\n",
       "Name: var_type, dtype: int64"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1 = vcf_to_dataframe('result.vcf')\n",
    "print (v1.var_type.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unknown    4342195\n",
      "indel        10772\n",
      "snp           7545\n",
      "Name: var_type, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "v2 = vcf_to_dataframe('mapped/raw.vcf')\n",
    "print (v2.var_type.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = 'bcftools filter -e \"QUAL<40\" -o filtered.vcf.gz mapped/calls.vcf'\n",
    "print (cmd)\n",
    "#tmp = subprocess.check_output(cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vcf_to_dataframe(vcf_file, quality=30):\n",
    "    \"\"\"Convert vcf to dataframe\"\"\"\n",
    "\n",
    "    import vcf\n",
    "    ext = os.path.splitext(vcf_file)[1]    \n",
    "    if ext == '.gz':\n",
    "        file = gzopen(vcf_file, \"rt\")\n",
    "    else:\n",
    "        file = open(vcf_file)\n",
    "    vcf_reader = vcf.Reader(file,'r')    \n",
    "    res=[]\n",
    "    cols = ['chrom','var_type','sub_type','start','end','REF','ALT','QUAL','DP']\n",
    "    i=0\n",
    "    for rec in vcf_reader:\n",
    "        #if i>10:\n",
    "        #    break\n",
    "        x = [rec.CHROM, rec.var_type, rec.var_subtype, rec.start, rec.end, rec.REF, str(rec.ALT[0]),\n",
    "            rec.QUAL, rec.INFO['DP']]\n",
    "        #print (rec.__dict__)\n",
    "        #print (rec.INFO.keys())\n",
    "        #for call in rec.samples:\n",
    "        #    print (call.sample, call.data, rec.genotype(call.sample)) \n",
    "        res.append(x)\n",
    "        #print (x)    \n",
    "        #i+=1\n",
    "    res = pd.DataFrame(res,columns=cols)    \n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "snp    2756\n",
       "Name: var_type, dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdf = vcf_to_dataframe('mapped/filtered.vcf.gz')\n",
    "vdf.var_type.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make fasta alignment from filtered VCF sites with multiple samples\n",
    "\n",
    "https://github.com/JosephCrispell/GeneralTools/blob/master/ProcessingPipeline/CreateFastaWithReferenceFromFiltered_28-06-17.pl\n",
    "\n",
    "\"The allele frequencies at each position in the aligned (against reference) sequence from each isolate were examined. For a haploid organism these frequencies are expected to be either 0 or 1, with some random variation expected from sequencing errors (Sobkowiak et al., 2018). A heterozygous site was defined as one where the allele frequencies were >0.05 and <0.95. Four cattle-derived sequences that had more than 150 heterozygous sites, and allele frequencies that were clustered and non-random (data not shown), were removed. In addition, 26 badger-derived and 16 cattle-derived M. bovis sequences were removed because of suspected errors in the metadata (Appendix 1: Investigating isolate metadata discrepancies).\n",
    "\n",
    "For the sequences from the remaining isolates (204 badger- and 169 cattle-derived isolates), alleles were called at each variant position if they had mapping quality â¥30, high-quality base depth â¥4 (applied to reverse and forward reads separately), read depth â¥30, and allele support â¥0.95. For any site that failed these criteria, if the allele called had been observed in a different isolate that had passed, a second round of filtering was conducted using a high-quality base depth of 5 (total across forward and reverse reads) and the same allele support. As recombination is thought to be extremely rare for mycobacteria (Namouchi et al., 2012), variants in close proximity could indicate a region that is difficult to sequence or under high selection. To avoid calling variants in these regions, variant positions within 10 bp of one another were removed. Following filtering, sequences from 11 badger and 10 cattle isolates that had insufficient coverage (<95%) of the variant positions were removed. Once the alignment was generated, sites with a consistency index less than 1, generally considered homoplasies (Farris, 1989), were removed (n = 4, of 14,991 sites) using HomoplasyFinder (v0.0.0.9; Crispell et al., 2019; RRID: SCR_017300). All the scripts necessary for the processing of the WGS data are freely available online.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta_alignment_from_vcf(vcf_file, ref, callback=None):\n",
    "    \"\"\"Get a fasta alignment for all snp sites in a multi sample\n",
    "    vcf file, including the reference sequence\"\"\"\n",
    "    \n",
    "    from pyfaidx import Fasta\n",
    "    from pyfaidx import FastaVariant\n",
    "    #index vcf\n",
    "    cmd = 'tabix -p vcf -f {i}'.format(i=vcf_file)\n",
    "    tmp = subprocess.check_output(cmd,shell=True)\n",
    "    #get samples?\n",
    "    import vcf\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    samples = vcf_reader.samples\n",
    "    print ('%s samples' %len(samples))\n",
    "    result = []\n",
    "    \n",
    "    #reference sequence\n",
    "    reference = Fasta(ref)\n",
    "    chrom = list(reference.keys())[0]\n",
    "\n",
    "    #get the set of all sites first\n",
    "    sites=[]\n",
    "    for sample in samples:\n",
    "        #print (sample)\n",
    "        variant = FastaVariant(ref, vcf_file, \n",
    "                                 sample=sample, het=True, hom=True)\n",
    "        pos = list(variant[chrom].variant_sites)\n",
    "        sites.extend(pos)\n",
    "        #print (sample)\n",
    "        #print (pos[:20])\n",
    "    sites = sorted(set(sites))\n",
    "    print ('using %s sites' %len(sites))\n",
    "    if callback != None:\n",
    "        callback('using %s sites' %len(sites))\n",
    "    #get reference sequence for site positions\n",
    "    refseq=[]\n",
    "    for p in sites:\n",
    "        refseq.append(reference[chrom][p-1].seq)\n",
    "    refseq = ''.join(refseq)\n",
    "    #print (refseq)\n",
    "    refrec = SeqRecord(Seq(refseq),id='ref')\n",
    "    result.append(refrec) \n",
    "\n",
    "    sites_matrix = {}\n",
    "    #iterate over variants in each sample\n",
    "    for sample in samples:        \n",
    "        seq=[]\n",
    "        variant = FastaVariant(ref, vcf_file, \n",
    "                                 sample=sample, het=True, hom=True)     \n",
    "        #for p in variant[chrom].variant_sites:\n",
    "        for p in sites:        \n",
    "            rec = variant[chrom][p-1:p]\n",
    "            #print (p,rec)\n",
    "            seq.append(rec.seq)\n",
    "        seq = ''.join(seq)\n",
    "        #print (seq)\n",
    "        seqrec = SeqRecord(Seq(seq),id=sample)\n",
    "        result.append(seqrec)\n",
    "        sites_matrix[sample] = list(seqrec)\n",
    "    df = pd.DataFrame(sites_matrix)\n",
    "    df.index=sites    \n",
    "    return result, df\n",
    "\n",
    "st=time.time()\n",
    "result = fasta_alignment_from_vcf('mapped/filtered.vcf.gz', app.ref_genome)\n",
    "SeqIO.write(result, 'variants.fa', 'fasta')\n",
    "print (time.time()-st)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogeny from MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS -f a -N 10 -T 8 -m GTRCAT -V -p 81287182 -x 18745047 -n variants -s variants.fa\n"
     ]
    }
   ],
   "source": [
    "def run_RAXML(infile, name='variants', threads=8):\n",
    "    \"\"\"Run Raxml pthreads\"\"\"\n",
    "\n",
    "    bootstraps = 10\n",
    "    model = 'GTRCAT'\n",
    "    s1 = random.randint(0,1e8)\n",
    "    s2 = random.randint(0,1e8)\n",
    "    \n",
    "    files = glob.glob('RAxML_*')\n",
    "    for f in files:\n",
    "        os.remove(f)\n",
    "    cmd = 'raxmlHPC-PTHREADS -f a -N {nb} -T {t} -m {m} -V -p {s1} -x {s2} -n {n} -s {i}'.format(t=threads,nb=bootstraps,n=name,i=infile,s1=s1,s2=s2,m=model)\n",
    "    print (cmd)\n",
    "    tmp = subprocess.check_output(cmd, shell=True)\n",
    "    return\n",
    "    \n",
    "run_RAXML('variants.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_tree(filename, labelmap=None):\n",
    "    \"\"\"Draw a tree \"\"\"\n",
    "    \n",
    "    from ete3 import Tree, NodeStyle, TreeStyle\n",
    "    t = Tree(filename)\n",
    "    t.set_outgroup('ref')    \n",
    "    if labelmap != None:\n",
    "        trees.set_tiplabels(t,labelmap)\n",
    "    trees.format_nodes(t)\n",
    "    ts = TreeStyle()\n",
    "    ts.scale=1000\n",
    "    t.render(\"%%inline\", tree_style=ts)\n",
    "    t.render(\"tree.png\", tree_style=ts)\n",
    "    return t\n",
    "\n",
    "sra['filename'] = sra.Run.apply(lambda x: 'mapped/'+x+'.bam')\n",
    "print (sra[:3])\n",
    "labelmap = dict(zip(sra.filename,sra.geo_loc_name_country))\n",
    "#print (labelmap)\n",
    "t = create_tree('RAxML_bipartitions.variants', labelmap)\n",
    "t.render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "tree=Phylo.read('RAxML_bestTree.variants','newick')\n",
    "Phylo.draw(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consequence calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = 'bcftools csq -f Mbovis_AF212297.fa -g  Mbovis_csq_format.gff col2_mbovis.vcf -Ot -o col2_mbovis.csq.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
