{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Wicklow isolates Pathogenie vs JC pipeline results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os,shutil,subprocess\n",
    "import glob\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 200)\n",
    "import pylab as plt\n",
    "import matplotlib as mpl\n",
    "from Bio import SeqIO\n",
    "sys.path.append('pathogenie')\n",
    "from snpgenie import tools, aligners, app, trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqID</th>\n",
       "      <th>Forward</th>\n",
       "      <th>Reverse</th>\n",
       "      <th>NumberMapped</th>\n",
       "      <th>MappingProp</th>\n",
       "      <th>SequencingDate</th>\n",
       "      <th>County</th>\n",
       "      <th>VCF</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>AverageDepth</th>\n",
       "      <th>Aliquot</th>\n",
       "      <th>Species</th>\n",
       "      <th>Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1034</td>\n",
       "      <td>1034_S91_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>1034_S91_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>902962</td>\n",
       "      <td>0.999386</td>\n",
       "      <td>24-09-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>1034_1.vcf.gz</td>\n",
       "      <td>0.900721</td>\n",
       "      <td>27.709380</td>\n",
       "      <td>TB19-001034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1143</td>\n",
       "      <td>1143_S21_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>1143_S21_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>4083983</td>\n",
       "      <td>0.997583</td>\n",
       "      <td>29-07-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>1143_1.vcf.gz</td>\n",
       "      <td>0.988193</td>\n",
       "      <td>122.031793</td>\n",
       "      <td>TB19-001143</td>\n",
       "      <td>Cow</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1147</td>\n",
       "      <td>1147_S91_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>1147_S91_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>2544407</td>\n",
       "      <td>0.998363</td>\n",
       "      <td>29-07-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>1147_2.vcf.gz</td>\n",
       "      <td>0.977731</td>\n",
       "      <td>76.976549</td>\n",
       "      <td>TB19-001147</td>\n",
       "      <td>Cow</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1212</td>\n",
       "      <td>1212_S58_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>1212_S58_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>1944433</td>\n",
       "      <td>0.702596</td>\n",
       "      <td>29-07-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>1212_3.vcf.gz</td>\n",
       "      <td>0.969578</td>\n",
       "      <td>52.156617</td>\n",
       "      <td>TB19-001212</td>\n",
       "      <td>Cow</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-11594</td>\n",
       "      <td>13-11594_S85_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>13-11594_S85_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>2988951</td>\n",
       "      <td>0.999445</td>\n",
       "      <td>15-03-19</td>\n",
       "      <td>Wicklow</td>\n",
       "      <td>13-11594_2.vcf.gz</td>\n",
       "      <td>0.984226</td>\n",
       "      <td>88.893592</td>\n",
       "      <td>TB14-011594</td>\n",
       "      <td>Deer</td>\n",
       "      <td>2014-11-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>59-MBovis</td>\n",
       "      <td>59-MBovis_S33_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>59-MBovis_S33_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>3450574</td>\n",
       "      <td>0.997713</td>\n",
       "      <td>09-07-18</td>\n",
       "      <td>Wicklow</td>\n",
       "      <td>59-MBovis_27.vcf.gz</td>\n",
       "      <td>0.992081</td>\n",
       "      <td>105.080641</td>\n",
       "      <td>TB15-005018</td>\n",
       "      <td>Cow</td>\n",
       "      <td>2015-06-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>7-10423</td>\n",
       "      <td>7-10423_S73_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>7-10423_S73_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>3444775</td>\n",
       "      <td>0.999483</td>\n",
       "      <td>15-03-19</td>\n",
       "      <td>Wicklow</td>\n",
       "      <td>7-10423_27.vcf.gz</td>\n",
       "      <td>0.981583</td>\n",
       "      <td>102.215471</td>\n",
       "      <td>TB14-010423</td>\n",
       "      <td>Deer</td>\n",
       "      <td>2014-10-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>717</td>\n",
       "      <td>717_S28_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>717_S28_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>5103197</td>\n",
       "      <td>0.998784</td>\n",
       "      <td>29-07-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>717_44.vcf.gz</td>\n",
       "      <td>0.988343</td>\n",
       "      <td>150.909946</td>\n",
       "      <td>TB19-000717</td>\n",
       "      <td>Badger</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>724p</td>\n",
       "      <td>724p_S44_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>724p_S44_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>3800780</td>\n",
       "      <td>0.998031</td>\n",
       "      <td>29-07-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>724p_45.vcf.gz</td>\n",
       "      <td>0.986323</td>\n",
       "      <td>114.284870</td>\n",
       "      <td>TB19-000724</td>\n",
       "      <td>Badger</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>973</td>\n",
       "      <td>973_S53_L001-4_R1_001.fastq.gz</td>\n",
       "      <td>973_S53_L001-4_R2_001.fastq.gz</td>\n",
       "      <td>4712858</td>\n",
       "      <td>0.933564</td>\n",
       "      <td>29-07-19</td>\n",
       "      <td>Monaghan</td>\n",
       "      <td>973_46.vcf.gz</td>\n",
       "      <td>0.991621</td>\n",
       "      <td>137.874159</td>\n",
       "      <td>TB19-000973</td>\n",
       "      <td>Cow</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>218 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SeqID                               Forward                               Reverse  NumberMapped  MappingProp SequencingDate    County                  VCF  Coverage  AverageDepth  \\\n",
       "0         1034       1034_S91_L001-4_R1_001.fastq.gz       1034_S91_L001-4_R2_001.fastq.gz        902962     0.999386       24-09-19  Monaghan        1034_1.vcf.gz  0.900721     27.709380   \n",
       "1         1143       1143_S21_L001-4_R1_001.fastq.gz       1143_S21_L001-4_R2_001.fastq.gz       4083983     0.997583       29-07-19  Monaghan        1143_1.vcf.gz  0.988193    122.031793   \n",
       "2         1147       1147_S91_L001-4_R1_001.fastq.gz       1147_S91_L001-4_R2_001.fastq.gz       2544407     0.998363       29-07-19  Monaghan        1147_2.vcf.gz  0.977731     76.976549   \n",
       "3         1212       1212_S58_L001-4_R1_001.fastq.gz       1212_S58_L001-4_R2_001.fastq.gz       1944433     0.702596       29-07-19  Monaghan        1212_3.vcf.gz  0.969578     52.156617   \n",
       "4     13-11594   13-11594_S85_L001-4_R1_001.fastq.gz   13-11594_S85_L001-4_R2_001.fastq.gz       2988951     0.999445       15-03-19   Wicklow    13-11594_2.vcf.gz  0.984226     88.893592   \n",
       "..         ...                                   ...                                   ...           ...          ...            ...       ...                  ...       ...           ...   \n",
       "213  59-MBovis  59-MBovis_S33_L001-4_R1_001.fastq.gz  59-MBovis_S33_L001-4_R2_001.fastq.gz       3450574     0.997713       09-07-18   Wicklow  59-MBovis_27.vcf.gz  0.992081    105.080641   \n",
       "214    7-10423    7-10423_S73_L001-4_R1_001.fastq.gz    7-10423_S73_L001-4_R2_001.fastq.gz       3444775     0.999483       15-03-19   Wicklow    7-10423_27.vcf.gz  0.981583    102.215471   \n",
       "215        717        717_S28_L001-4_R1_001.fastq.gz        717_S28_L001-4_R2_001.fastq.gz       5103197     0.998784       29-07-19  Monaghan        717_44.vcf.gz  0.988343    150.909946   \n",
       "216       724p       724p_S44_L001-4_R1_001.fastq.gz       724p_S44_L001-4_R2_001.fastq.gz       3800780     0.998031       29-07-19  Monaghan       724p_45.vcf.gz  0.986323    114.284870   \n",
       "217        973        973_S53_L001-4_R1_001.fastq.gz        973_S53_L001-4_R2_001.fastq.gz       4712858     0.933564       29-07-19  Monaghan        973_46.vcf.gz  0.991621    137.874159   \n",
       "\n",
       "         Aliquot Species        Date  \n",
       "0    TB19-001034     NaN         NaN  \n",
       "1    TB19-001143     Cow         NaN  \n",
       "2    TB19-001147     Cow         NaN  \n",
       "3    TB19-001212     Cow         NaN  \n",
       "4    TB14-011594    Deer  2014-11-12  \n",
       "..           ...     ...         ...  \n",
       "213  TB15-005018     Cow  2015-06-11  \n",
       "214  TB14-010423    Deer  2014-10-23  \n",
       "215  TB19-000717  Badger         NaN  \n",
       "216  TB19-000724  Badger         NaN  \n",
       "217  TB19-000973     Cow         NaN  \n",
       "\n",
       "[218 rows x 13 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv('/storage/btbgenie/JoeROIData/Mbovis/sampleInformation_NoAnimalIDs_18-12-19.csv')\n",
    "meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run workflow "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(app)\n",
    "reload(tools)\n",
    "args = {'threads':10, 'outdir': '/storage/btbgenie/wicklow_results', 'labelsep':'_',\n",
    "        'input':['/storage/btbgenie/JoeROIData/Mbovis/Wicklow/Fastqs_07-01-18/',\n",
    "                 '/storage/btbgenie/JoeROIData/Mbovis/Wicklow/Fastqs_15-03-19/',\n",
    "                 '/storage/btbgenie/JoeROIData/Mbovis/Wicklow/Fastqs_09-07-18/'],       \n",
    "         'reference': None, 'overwrite':False,\n",
    "         'trim': False,\n",
    "         #'filters' : 'QUAL>=30 && FORMAT/DP>=30 && DP4>=4 && MQ>35',\n",
    "         #'custom_filters': True,\n",
    "         'buildtree':True}\n",
    "W = app.WorkFlow(**args)\n",
    "st = W.setup()\n",
    "W.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare sites matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "smat = pd.read_csv('/storage/btbgenie/wicklow_results/core.txt',sep=' ',index_col=0)\n",
    "smat.index\n",
    "joe_sites = pd.read_csv('/storage/btbgenie/JoeROIData/Mbovis/Wicklow/FASTA/fastaPositions_Prox-10_18-12-2019.txt')\n",
    "#print (joe_sites)\n",
    "plt.figure(figsize=(6,6))\n",
    "import matplotlib_venn\n",
    "mpl.rc('text', usetex=True)\n",
    "mpl.rc('font',size=14)\n",
    "sets=[set(joe_sites.Position),set(smat.index)]\n",
    "matplotlib_venn.venn2(sets,set_labels=['Joe','snpgenie'],set_colors=['#4d88ff','#ff3333'])\n",
    "ax=matplotlib_venn.venn2_circles(sets,linewidth=1)\n",
    "plt.title('SNP site overlap',fontsize=20)\n",
    "plt.savefig('snp_overlap_joe.png',dpi=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.071429    1798\n",
       "0.214286     881\n",
       "0.142857     403\n",
       "0.857143     153\n",
       "0.785714      77\n",
       "1.000000      64\n",
       "0.285714      36\n",
       "0.500000      21\n",
       "0.571429      10\n",
       "0.714286       6\n",
       "0.642857       3\n",
       "0.428571       3\n",
       "0.357143       2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#frequencies of sites\n",
    "csqmat = pd.read_csv('../test_results/csq.matrix').set_index(['gene','aa'])\n",
    "sitefreqs = (csqmat.sum(1)/len(csqmat.columns)).sort_values(ascending=False)\n",
    "sitefreqs.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqID                                        1034\n",
      "Forward           1034_S91_L001-4_R1_001.fastq.gz\n",
      "Reverse           1034_S91_L001-4_R2_001.fastq.gz\n",
      "NumberMapped                               902962\n",
      "MappingProp                              0.999386\n",
      "SequencingDate                           24-09-19\n",
      "County                                   Monaghan\n",
      "VCF                                 1034_1.vcf.gz\n",
      "Coverage                                 0.900721\n",
      "AverageDepth                              27.7094\n",
      "Aliquot                               TB19-001034\n",
      "Species                                       NaN\n",
      "Date                                          NaN\n",
      "name                       1034_S91_L001-4_R1_001\n",
      "Name: 0, dtype: object\n"
     ]
    }
   ],
   "source": [
    "#print(W.fastq_table)\n",
    "meta['name'] = meta.Forward.apply(lambda x: x.split('.')[0])\n",
    "print(meta.loc[0])\n",
    "s = pd.merge(W.fastq_table,meta,on='name')\n",
    "labelmap = dict(zip(s['sample'],s.County))\n",
    "#print (labelmap)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "NewickError",
     "evalue": "Unexisting tree file or Malformed newick tree structure.\nYou may want to check other newick loading flags like 'format' or 'quoted_node_names'.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNewickError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-986361586e69>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mreload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrees\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mt\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrees\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_tree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/storage/btbgenie/wicklow_results/RAxML_bipartitions.variants'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'ref'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlabelmap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlabelmap\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrender\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%%inline\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtree_style\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/gitprojects/snpgenie/snpgenie/trees.py\u001b[0m in \u001b[0;36mcreate_tree\u001b[0;34m(filename, ref, labelmap, colormap)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0mete3\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mTree\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTreeStyle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTextFace\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 117\u001b[0;31m     \u001b[0mt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    118\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mref\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    119\u001b[0m         \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_outgroup\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/gitprojects/ete/ete3/coretype/tree.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, newick, format, dist, support, name, quoted_node_names)\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mnewick\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    209\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 210\u001b[0;31m             read_newick(newick, root_node = self, format=format,\n\u001b[0m\u001b[1;32m    211\u001b[0m                         quoted_names=quoted_node_names)\n\u001b[1;32m    212\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/gitprojects/ete/ete3/parser/newick.py\u001b[0m in \u001b[0;36mread_newick\u001b[0;34m(newick, root_node, format, quoted_names)\u001b[0m\n\u001b[1;32m    260\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0m_read_newick_from_string\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mroot_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatcher\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquoted_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnw\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'('\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnw\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mendswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m';'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mNewickError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Unexisting tree file or Malformed newick tree structure.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0m_read_newick_from_string\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mroot_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatcher\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquoted_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNewickError\u001b[0m: Unexisting tree file or Malformed newick tree structure.\nYou may want to check other newick loading flags like 'format' or 'quoted_node_names'."
     ]
    }
   ],
   "source": [
    "reload(trees)\n",
    "t,ts = trees.create_tree('/storage/btbgenie/wicklow_results/RAxML_bipartitions.variants', ref='ref',labelmap=labelmap)\n",
    "t.render(\"%%inline\",tree_style=ts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(trees)\n",
    "t = trees.create_tree('/storage/btbgenie/JoeROIData/Mbovis/Wicklow/mlTree_WICKLOW_18-12-2019.tree')#, labelmap)\n",
    "t.render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "csqmat = pd.read_csv('')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
